File: CS/ICustomTypeDescriptor/ViewModels/MainViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Windows.Input;
using DevExpress.Mvvm;
using DevExpress.Mvvm.Native;

namespace ICustomTypeDescriptor.ViewModels {
    public class MainViewModel : ViewModelBase {
        public ObservableCollection<Item> Items { get => GetProperty(() => Items); set => SetProperty(() => Items, value); }

        static MainViewModel() {
            ItemTypeDescriptor.CustomFields.Add(new ItemPropertyDescriptor(new CustomField("Name", typeof(string))));
            ItemTypeDescriptor.CustomFields.Add(
                new ItemPropertyDescriptor(new CustomField("CreatedAt", typeof(DateTime))));
        }

        public MainViewModel() {
            Items = new ObservableCollection<Item>(Enumerable.Range(0, 10).Select(i => new Item { Id = i }));
            Items.ForEach(x => x["Name"] = $"Item {Items.IndexOf(x)}");
            Items.ForEach(x => x["CreatedAt"] = DateTime.Now.AddDays(Items.IndexOf(x)));

            AddColumnCommand = new DelegateCommand(AddColumn);
        }

        public void AddColumn() {
            var fieldCount = ItemTypeDescriptor.CustomFields.Count;
            ItemTypeDescriptor.CustomFields.Add(
                new ItemPropertyDescriptor(new CustomField($"Value {fieldCount}", typeof(int))));
            Items.ForEach(x => x[$"Value {fieldCount}"] = Items.IndexOf(x) * fieldCount);
        }

        public ICommand AddColumnCommand { get; }
    }

    [TypeDescriptionProvider(typeof(ItemDescriptionProvider))]
    public class Item : BindableBase {
        private readonly Dictionary<string, object> _customFieldValues = new();

        public int Id { get => GetProperty(() => Id); set => SetProperty(() => Id, value); }

        public object this[string fieldName] {
            get {
                _customFieldValues.TryGetValue(fieldName, out var value);
                return value;
            }

            set {
                _customFieldValues[fieldName] = value;
                RaisePropertyChanged(fieldName);
            }
        }
    }
}

File: VB/ICustomTypeDescriptor/ViewModels/MainViewModel.vb
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports DevExpress.Mvvm.CodeGenerators
Imports DevExpress.Mvvm.Native

Namespace ICustomTypeDescriptor.ViewModels

    <GenerateViewModel>
    Public Partial Class MainViewModel

        <GenerateProperty>
        Private _items As ObservableCollection(Of Item)

        Shared Sub New()
            ItemTypeDescriptor.CustomFields.Add(New ItemPropertyDescriptor(New CustomField("Name", GetType(String))))
            ItemTypeDescriptor.CustomFields.Add(New ItemPropertyDescriptor(New CustomField("CreatedAt", GetType(Date))))
        End Sub

        Public Sub New()
            _items = New ObservableCollection(Of Item)(Enumerable.Range(0, 10).[Select](Function(i) New Item With {.Id = i}))
            _items.ForEach(Sub(x) x("Name") = $"Item {_items.IndexOf(x)}")
            _items.ForEach(Sub(x) x("CreatedAt") = Date.Now.AddDays(_items.IndexOf(x)))
        End Sub

        <GenerateCommand>
        Public Sub AddColumn()
            Dim fieldCount = ItemTypeDescriptor.CustomFields.Count
            ItemTypeDescriptor.CustomFields.Add(New ItemPropertyDescriptor(New CustomField($"Value {fieldCount}", GetType(Integer))))
            _items.ForEach(Sub(x) x($"Value {fieldCount}") = _items.IndexOf(x) * fieldCount)
        End Sub
    End Class

    <GenerateViewModel>
    <TypeDescriptionProvider(GetType(ItemDescriptionProvider))>
    Public Partial Class Item
        Implements INotifyPropertyChanged

        Private ReadOnly _customFieldValues As Dictionary(Of String, Object) = New(_, _)()

        <GenerateProperty>
        Private _id As Integer

        Default Public Property Item(ByVal fieldName As String) As Object
            Get
                Dim value As Object = Nothing
                _customFieldValues.TryGetValue(fieldName, value)
                Return value
            End Get

            Set(ByVal value As Object)
                _customFieldValues(fieldName) = value
                OnPropertyChanged(fieldName)
            End Set
        End Property

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Overridable Sub OnPropertyChanged(<CallerMemberName> ByVal Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class
End Namespace
